Projection Optimization Rule

berkaysynnada opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

berkaysynnada committed:
Draft PR is moved to new repo

mustafasrepo committed:
Add new reproducer test

mustafasrepo committed:
Simplifications

mustafasrepo committed:
Simplifications

mustafasrepo committed:
Simplifications

mustafasrepo committed:
Simplifications

mustafasrepo committed:
Simpifications

mustafasrepo committed:
Simplifications

mustafasrepo committed:
Minor changes

berkaysynnada committed:
tpch fails

metesynnada committed:
Delete docs.yaml

mustafasrepo committed:
Merge pull request #5 from synnada-ai/ci-action-fixing

Fix deploying DataFusion site error

berkaysynnada committed:
All tests pass, ready to be cleaned-up

berkaysynnada committed:
Merge branch 'main' into feature/optimize-projections

mustafasrepo committed:
Merge branch 'apache:main' into main

berkaysynnada committed:
Prevent projection removals causing uncertain column swaps

mustafasrepo committed:
Merge branch 'apache:main' into main

berkaysynnada committed:
Minor changes

berkaysynnada committed:
Merge branch 'main' into feature/optimize-projections

berkaysynnada committed:
Minor changes

berkaysynnada committed:
Update optimize_projections.rs

metesynnada committed:
Mini Review before unwraps - Part 1

metesynnada committed:
Continue on review - Part 2

metesynnada commented on datafusion/core/src/physical_optimizer/optimize_projections.rs:
```diff
@@ -0,0 +1,4955 @@
+        Ok(self)
+    }
+
+    fn try_insert_below_interleave(mut self) -> Result<ProjectionOptimizer> {
```
Remove recursive clones from all join related insertions

metesynnada committed:
Adding comments - Review Part 3

berkaysynnada committed:
Clone's and unwrap's are removed

berkaysynnada committed:
Update optimize_projections.rs

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
add yaml

berkaysynnada committed:
projection update

berkaysynnada committed:
Pushdown through limits

berkaysynnada committed:
fix clippy

berkaysynnada committed:
ProjectionMapping refactor

berkaysynnada committed:
Sync all joins

berkaysynnada committed:
Adapt test changes

berkaysynnada committed:
Remove duplications

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Update optimize_projections.rs

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Fix after merge

berkaysynnada committed:
Inherit hashjoin projection

mustafasrepo commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -4029,13 +4032,14 @@ Projection: aggregate_test_100.c3, SUM(aggregate_test_100.c9) ORDER BY [aggregat
---------BoundedWindowAggExec: wdw=[SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 DESC NULLS FIRST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW: Ok(Field { name: "SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 DESC NULLS FIRST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW", data_type: UInt64, nullable: true, dict_i ...
-----------SortExec: expr=[c3@0 + c4@1 DESC]
-------------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/testing/data/csv/aggregate_test_100.csv]]}, projection=[c3, c4, c9], has_header=true
+----ProjectionExec: expr=[c3@0 as c3, sum1@3 as sum1, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@4 as SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW]
```
This Projection seems unnecessary.

berkaysynnada committed:
Minor changes

mustafasrepo commented on datafusion/core/src/physical_optimizer/optimize_projections.rs:
```diff
@@ -0,0 +1,5817 @@
+            "  ProjectionExec: expr=[c@0 + x@1 as sum]", 
+            "    SortExec: expr=[c@0 ASC,x@1 ASC]", 
+            "      ProjectionExec: expr=[c@2 as c, a@0 as x]",
+            "        CsvExec: file_groups={1 group: [[x]]}, projection=[a, b, c, d, e], has_header=false"];
```
Projection should be pushed down to Csv for this test

mustafasrepo commented on datafusion/core/tests/sql/explain_analyze.rs:
```diff
@@ -656,17 +656,18 @@ async fn test_physical_plan_display_indent_multi_children() {
-    	"            CsvExec: file_groups={1 group: [[ARROW_TEST_DATA/csv/aggregate_test_100.csv]]}, projection=[c1], has_header=true",
+        "CoalesceBatchesExec: target_batch_size=4096", 
+        "  ProjectionExec: expr=[c1@0 as c1]", 
+        "    HashJoinExec: mode=Partitioned, join_type=Inner, on=[(c1@0, c2@0)]", 
```
Previous plan is better. Builtin projection should be used

mustafasrepo commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -2873,28 +2873,29 @@ Sort: s.sn ASC NULLS LAST
 ------ProjectionExec: expr=[zip_code@2 as zip_code, country@3 as country, sn@4 as sn, ts@5 as ts, currency@6 as currency, sn@0 as sn, amount@1 as amount]
 --------CoalesceBatchesExec: target_batch_size=8192
-----------HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(currency@2, currency@4)], filter=ts@0 >= ts@1, projection=[sn@0, amount@3, zip_code@4, country@5, sn@6, ts@7, currency@8]
-------------MemoryExec: partitions=1, partition_sizes=[1]
-------------MemoryExec: partitions=1, partition_sizes=[1]
+----------ProjectionExec: expr=[sn@0 as sn, amount@3 as amount, zip_code@4 as zip_code, country@5 as country, sn@6 as sn, ts@7 as ts, currency@8 as currency]
```
Projection + Coalesce + Projection stack should replaced with Projection + Coalesce stack.

mustafasrepo replied:
Add this test case also as unit test in `optimize_projections.rs` file

mustafasrepo commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -3869,14 +3870,15 @@ Projection: amount_usd
---------ProjectionExec: expr=[a@0 as a, d@1 as d, ROW_NUMBER() ORDER BY [r.a ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@2 as row_n]
-----------BoundedWindowAggExec: wdw=[ROW_NUMBER() ORDER BY [r.a ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW: Ok(Field { name: "ROW_NUMBER() ORDER BY [r.a ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW", data_type: UInt64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(Int32(NU ...
-------------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/window_2.csv]]}, projection=[a, d], output_ordering=[a@0 ASC NULLS LAST], has_header=true
+------ProjectionExec: expr=[a@0 as a, d@1 as d, row_n@4 as row_n]
```
This projection might be unnecessary. Since aggregate kind of works like projection. We do not need to prune columns at its input.

mustafasrepo commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -4028,13 +4030,14 @@ Projection: lhs.c, rhs.c, lhs.sum1, rhs.sum1
---------AggregateExec: mode=Single, gby=[c@1 as c, b@0 as b], aggr=[SUM(multiple_ordered_table_with_pk.d)], ordering_mode=PartiallySorted([0])
-----------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/window_2.csv]]}, projection=[b, c, d], output_ordering=[c@1 ASC NULLS LAST], has_header=true
-------ProjectionExec: expr=[c@0 as c, b@1 as b, SUM(multiple_ordered_table_with_pk.d)@2 as sum1]
---------AggregateExec: mode=Single, gby=[c@1 as c, b@0 as b], aggr=[SUM(multiple_ordered_table_with_pk.d)], ordering_mode=PartiallySorted([0])
-----------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/window_2.csv]]}, projection=[b, c, d], output_ordering=[c@1 ASC NULLS LAST], has_header=true
+----ProjectionExec: expr=[c@0 as c, sum1@2 as sum1, c@3 as c, sum1@5 as sum1]
```
Same pattern Projection + SchemaPreservingOperator + Projection stack. This should be replaced with Projection + SchemaPreservingOperator (as long as  projection above is not column pruning projection)

mustafasrepo commented on datafusion/sqllogictest/test_files/join_disable_repartition_joins.slt:
```diff
@@ -101,12 +102,13 @@ GlobalLimitExec: skip=0, fetch=10
-----------CoalesceBatchesExec: target_batch_size=8192
-------------FilterExec: d@3 = 3
---------------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
-----------------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/window_2.csv]]}, projection=[a, b, c, d], output_ordering=[a@0 ASC NULLS LAST, b@1 ASC NULLS LAST, c@2 ASC NULLS LAST], has_header=true
+--------ProjectionExec: expr=[a@0 as a, b@1 as b]
```
Same pattern

mustafasrepo commented on datafusion/sqllogictest/test_files/joins.slt:
```diff
@@ -1326,15 +1326,16 @@ Aggregate: groupBy=[[join_t1.t1_id]], aggr=[[]]
---------RepartitionExec: partitioning=Hash([t2_id@0], 2), input_partitions=2
-----------RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1
-------------MemoryExec: partitions=1, partition_sizes=[1]
+----ProjectionExec: expr=[t1_id@0 as t1_id]
```
I think, pruned columns at the input of Coalesce is beneficial in terms of memory usage. Hence, these changes are in good direction. Similar kind of changes exist for this reason in `.slt` files. I wonder what others think?

mustafasrepo commented on datafusion/sqllogictest/test_files/joins.slt:
```diff
@@ -1353,15 +1354,16 @@ physical_plan
-----------RepartitionExec: partitioning=Hash([t2_id@0], 2), input_partitions=2
-------------RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1
---------------MemoryExec: partitions=1, partition_sizes=[1]
+------ProjectionExec: expr=[t1_id@0 as t1_id]
```
Same above columns are pruned at the input of `CoalesceBatchesExec`.

mustafasrepo commented on datafusion/sqllogictest/test_files/joins.slt:
```diff
@@ -3407,14 +3420,15 @@ Projection: amount_usd
---------ProjectionExec: expr=[a@0 as a, d@1 as d, ROW_NUMBER() ORDER BY [r.a ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@2 as row_n]
-----------BoundedWindowAggExec: wdw=[ROW_NUMBER() ORDER BY [r.a ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW: Ok(Field { name: "ROW_NUMBER() ORDER BY [r.a ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW", data_type: UInt64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(Int32(NU ...
-------------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/window_2.csv]]}, projection=[a, d], output_ordering=[a@0 ASC NULLS LAST], has_header=true
+------ProjectionExec: expr=[a@0 as a, d@1 as d, row_n@4 as row_n]
```
Most of the changes in this file comes from the `CoalesceBatchesExec` behaviour mentioned above.

mustafasrepo commented on datafusion/sqllogictest/test_files/subquery.slt:
```diff
@@ -189,16 +189,16 @@ Projection: t1.t1_id, __scalar_sq_1.SUM(t2.t2_int) AS t2_sum
 --CoalesceBatchesExec: target_batch_size=2
-----HashJoinExec: mode=Partitioned, join_type=Right, on=[(t2_id@1, t1_id@0)], projection=[SUM(t2.t2_int)@0, t1_id@2]
-------ProjectionExec: expr=[SUM(t2.t2_int)@1 as SUM(t2.t2_int), t2_id@0 as t2_id]
+----ProjectionExec: expr=[SUM(t2.t2_int)@1 as SUM(t2.t2_int), t1_id@2 as t1_id]
+------HashJoinExec: mode=Partitioned, join_type=Right, on=[(t2_id@0, t1_id@0)]
```
This is good. Because, projection before hash doesn't prune table. Hence hash doesn't benefit from it. Reshuffling columns are delayed until it is necessary. However this query has afore mentioned unnecessary `Projection + Coalesce + Projection` stack problem.

mustafasrepo commented on datafusion/sqllogictest/test_files/subquery.slt:
```diff
@@ -223,16 +223,17 @@ Projection: t1.t1_id, __scalar_sq_1.SUM(t2.t2_int * Float64(1)) + Int64(1) AS t2
-------CoalesceBatchesExec: target_batch_size=2
---------RepartitionExec: partitioning=Hash([t1_id@0], 4), input_partitions=4
-----------MemoryExec: partitions=4, partition_sizes=[1, 0, 0, 0]
+----ProjectionExec: expr=[SUM(t2.t2_int * Float64(1)) + Int64(1)@0 as SUM(t2.t2_int * Float64(1)) + Int64(1), t1_id@2 as t1_id]
```
Same above

mustafasrepo commented on datafusion/sqllogictest/test_files/subquery.slt:
```diff
@@ -257,18 +258,19 @@ Projection: t1.t1_id, __scalar_sq_1.SUM(t2.t2_int) AS t2_sum
-----------------RepartitionExec: partitioning=Hash([t2_id@0, Utf8("a")@1], 4), input_partitions=4
-------------------AggregateExec: mode=Partial, gby=[t2_id@0 as t2_id, a as Utf8("a")], aggr=[SUM(t2.t2_int)], ordering_mode=PartiallySorted([1])
---------------------MemoryExec: partitions=4, partition_sizes=[1, 0, 0, 0]
+----ProjectionExec: expr=[t1_id@0 as t1_id, SUM(t2.t2_int)@1 as SUM(t2.t2_int)]
```
Same above

mustafasrepo commented on datafusion/sqllogictest/test_files/subquery.slt:
```diff
@@ -294,18 +296,18 @@ Projection: t1.t1_id, __scalar_sq_1.SUM(t2.t2_int) AS t2_sum
 --CoalesceBatchesExec: target_batch_size=2
-----HashJoinExec: mode=Partitioned, join_type=Right, on=[(t2_id@1, t1_id@0)], projection=[SUM(t2.t2_int)@0, t1_id@2]
-------ProjectionExec: expr=[SUM(t2.t2_int)@1 as SUM(t2.t2_int), t2_id@0 as t2_id]
+----ProjectionExec: expr=[SUM(t2.t2_int)@1 as SUM(t2.t2_int), t1_id@2 as t1_id]
```
Same above

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q1.slt.part:
```diff
@@ -56,8 +56,9 @@ SortPreservingMergeExec: [l_returnflag@0 ASC NULLS LAST,l_linestatus@1 ASC NULLS
 ----------------CoalesceBatchesExec: target_batch_size=8192
-------------------FilterExec: l_shipdate@6 <= 10471
---------------------CsvExec: file_groups={4 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:0..18561749], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:18561749..37123498], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:37123498..55685247], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineite ...
+------------------ProjectionExec: expr=[l_quantity@0 as l_quantity, l_extendedprice@1 as l_extendedprice, l_discount@2 as l_discount, l_tax@3 as l_tax, l_returnflag@4 as l_returnflag, l_linestatus@5 as l_linestatus]
```
This change is good

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q10.slt.part:
```diff
@@ -79,35 +79,38 @@ GlobalLimitExec: skip=0, fetch=10
---------------------------HashJoinExec: mode=Partitioned, join_type=Inner, on=[(o_orderkey@7, l_orderkey@0)], projection=[c_custkey@0, c_name@1, c_address@2, c_nationkey@3, c_phone@4, c_acctbal@5, c_comment@6, l_extendedprice@9, l_discount@10]
-----------------------------CoalesceBatchesExec: target_batch_size=8192
-------------------------------RepartitionExec: partitioning=Hash([o_orderkey@7], 4), input_partitions=4
+------------------ProjectionExec: expr=[c_custkey@0 as c_custkey, c_name@1 as c_name, c_address@2 as c_address, c_phone@4 as c_phone, c_acctbal@5 as c_acctbal, c_comment@6 as c_comment, l_extendedprice@7 as l_extendedprice, l_discount@8 as l_discount, n_name@10 as n_name]
```
These changes seems to be good.

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q11.slt.part:
```diff
@@ -84,51 +84,53 @@ GlobalLimitExec: skip=0, fetch=10
-----------------------CoalesceBatchesExec: target_batch_size=8192
-------------------------RepartitionExec: partitioning=Hash([n_nationkey@0], 4), input_partitions=4
---------------------------ProjectionExec: expr=[n_nationkey@0 as n_nationkey]
+--------------------ProjectionExec: expr=[ps_partkey@0 as ps_partkey, ps_availqty@1 as ps_availqty, ps_supplycost@2 as ps_supplycost]
```
These changes seems to be good.

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q12.slt.part:
```diff
@@ -67,16 +67,17 @@ SortPreservingMergeExec: [l_shipmode@0 ASC NULLS LAST]
-------------------CoalesceBatchesExec: target_batch_size=8192
---------------------RepartitionExec: partitioning=Hash([l_orderkey@0], 4), input_partitions=4
-----------------------ProjectionExec: expr=[l_orderkey@0 as l_orderkey, l_shipmode@4 as l_shipmode]
+----------------ProjectionExec: expr=[l_shipmode@1 as l_shipmode, o_orderpriority@3 as o_orderpriority]
```
These changes seems to be good.

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q13.slt.part:
```diff
@@ -65,17 +65,18 @@ GlobalLimitExec: skip=0, fetch=10
-------------------------CoalesceBatchesExec: target_batch_size=8192
---------------------------RepartitionExec: partitioning=Hash([o_custkey@1], 4), input_partitions=4
-----------------------------ProjectionExec: expr=[o_orderkey@0 as o_orderkey, o_custkey@1 as o_custkey]
+----------------------ProjectionExec: expr=[c_custkey@0 as c_custkey, o_orderkey@1 as o_orderkey]
```
These changes seems to be good.

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q14.slt.part:
```diff
@@ -46,17 +46,18 @@ ProjectionExec: expr=[100 * CAST(SUM(CASE WHEN part.p_type LIKE Utf8("PROMO%")
-------------CoalesceBatchesExec: target_batch_size=8192
---------------RepartitionExec: partitioning=Hash([l_partkey@0], 4), input_partitions=4
-----------------ProjectionExec: expr=[l_partkey@0 as l_partkey, l_extendedprice@1 as l_extendedprice, l_discount@2 as l_discount]
+----------ProjectionExec: expr=[l_extendedprice@1 as l_extendedprice, l_discount@2 as l_discount, p_type@4 as p_type]
```
These changes seems to be good.

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q15.slt.part:
```diff
@@ -74,38 +74,40 @@ physical_plan
-----------------------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/supplier.tbl]]}, projection=[s_suppkey, s_name, s_address, s_phone], has_header=false
-----------------ProjectionExec: expr=[l_suppkey@0 as supplier_no, SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount)@1 as total_revenue]
-------------------AggregateExec: mode=FinalPartitioned, gby=[l_suppkey@0 as l_suppkey], aggr=[SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount)]
+------ProjectionExec: expr=[s_suppkey@0 as s_suppkey, s_name@1 as s_name, s_address@2 as s_address, s_phone@3 as s_phone, total_revenue@4 as total_revenue]
```
These changes seems to be good.

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q16.slt.part:
```diff
@@ -83,20 +83,21 @@ GlobalLimitExec: skip=0, fetch=10
-------------------------------------------FilterExec: p_brand@1 != Brand#45 AND p_type@2 NOT LIKE MEDIUM POLISHED% AND Use p_size@3 IN (SET) ([Literal { value: Int32(49) }, Literal { value: Int32(14) }, Literal { value: Int32(23) }, Literal { value: Int32(45) }, Literal { value: Int32(19) }, Literal { value: Int32(3) }, Literal { value: Int32(36) }, Literal { value: Int32(9) }])
---------------------------------------------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
-----------------------------------------------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/part.tbl]]}, projection=[p_partkey, p_brand, p_type, p_size], has_header=false
+----------------------------------ProjectionExec: expr=[ps_suppkey@1 as ps_suppkey, p_brand@3 as p_brand, p_type@4 as p_type, p_size@5 as p_size]
```
These changes seems to be good.

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q17.slt.part:
```diff
@@ -56,25 +56,27 @@ ProjectionExec: expr=[CAST(SUM(lineitem.l_extendedprice)@0 AS Float64) / 7 as av
-------------------RepartitionExec: partitioning=Hash([l_partkey@0], 4), input_partitions=4
---------------------AggregateExec: mode=Partial, gby=[l_partkey@0 as l_partkey], aggr=[AVG(lineitem.l_quantity)]
-----------------------CsvExec: file_groups={4 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:0..18561749], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:18561749..37123498], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:37123498..55685247], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/linei ...
+----------ProjectionExec: expr=[l_extendedprice@1 as l_extendedprice]
```
These changes seems to be good.

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q18.slt.part:
```diff
@@ -76,23 +76,25 @@ SortPreservingMergeExec: [o_totalprice@4 DESC,o_orderdate@3 ASC NULLS LAST]
+------------------------RepartitionExec: partitioning=Hash([l_orderkey@0], 4), input_partitions=4
+--------------------------CsvExec: file_groups={4 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:0..18561749], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:18561749..37123498], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:37123498..55685247], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/l ...
+----------------CoalesceBatchesExec: target_batch_size=8192
+------------------ProjectionExec: expr=[l_orderkey@0 as l_orderkey]
```
These changes seems to be good.

mustafasrepo commented on datafusion/sqllogictest/test_files/tpch/q2.slt.part:
```diff
@@ -100,90 +100,95 @@ Limit: skip=0, fetch=10
---SortPreservingMergeExec: [s_acctbal@0 DESC,n_name@2 ASC NULLS LAST,s_name@1 ASC NULLS LAST,p_partkey@3 ASC NULLS LAST], fetch=10
-----SortExec: TopK(fetch=10), expr=[s_acctbal@0 DESC,n_name@2 ASC NULLS LAST,s_name@1 ASC NULLS LAST,p_partkey@3 ASC NULLS LAST]
-------ProjectionExec: expr=[s_acctbal@5 as s_acctbal, s_name@2 as s_name, n_name@7 as n_name, p_partkey@0 as p_partkey, p_mfgr@1 as p_mfgr, s_address@3 as s_address, s_phone@4 as s_phone, s_comment@6 as s_comment]
+ProjectionExec: expr=[s_acctbal@5 as s_acctbal, s_name@2 as s_name, n_name@7 as n_name, p_partkey@0 as p_partkey, p_mfgr@1 as p_mfgr, s_address@3 as s_address, s_phone@4 as s_phone, s_comment@6 as s_comment]
```
Same problem above, with more than one schema preserving operator in between

mustafasrepo commented on datafusion/sqllogictest/test_files/union.slt:
```diff
@@ -278,34 +278,35 @@ Union
-----------RepartitionExec: partitioning=Hash([id@0, name@1], 4), input_partitions=4
-------------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
---------------MemoryExec: partitions=1, partition_sizes=[1]
+------ProjectionExec: expr=[id@0 as id, name@1 as name]
```
Same problem mentioned before

mustafasrepo commented:
I have looked to the tests in this PR.
Generally changes are good and beneficial (according to my expectations)
However, there are 3 points need to be resolved

- builtin projection handling for `HashJoinExec`
- Collapsing `Projection` + `SchemaPreservingOperator` (n times) + `Projection (shuffler not pruning or calculating)` stack into `Projection` + `SchemaPreservingOperator` (n times)
- CsvExec source projection pushdown handling

berkaysynnada committed:
Project csv only with requirements

berkaysynnada committed:
to check diff

berkaysynnada committed:
too many open files!

berkaysynnada committed:
tests pass

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
get upstream changes

berkaysynnada committed:
test changes

berkaysynnada committed:
Unify hashjoins

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Addressing mete's latest feedback

mustafasrepo commented on datafusion/core/tests/sql/explain_analyze.rs:
```diff
@@ -656,17 +656,17 @@ async fn test_physical_plan_display_indent_multi_children() {
-    	"        RepartitionExec: partitioning=RoundRobinBatch(9000), input_partitions=1",
-    	"          ProjectionExec: expr=[c1@0 as c2]",
-    	"            CsvExec: file_groups={1 group: [[ARROW_TEST_DATA/csv/aggregate_test_100.csv]]}, projection=[c1], has_header=true",
+        "CoalesceBatchesExec: target_batch_size=4096",
```
This change comes from formatting difference. I think you can retract it to no clutter PR.

berkaysynnada committed:
Fixing the bug during benchmarks

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Update cte.slt

berkaysynnada committed:
tests added

berkaysynnada committed:
checked till try_remove

berkaysynnada committed:
Update optimize_projections.rs

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Update optimize_projections.rs

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
Why are the `ORDER BY ...` prints gone?

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -3420,9 +3420,9 @@ Sort: r.sn ASC NULLS LAST
-----ProjectionExec: expr=[sn@0 as sn, SUM(l.amount)@2 as SUM(l.amount), amount@1 as amount]
+ProjectionExec: expr=[sn@0 as sn, SUM(l.amount)@2 as SUM(l.amount), amount@1 as amount]
+--SortPreservingMergeExec: [sn@0 ASC NULLS LAST]
+----SortExec: expr=[sn@0 ASC NULLS LAST]
```
When sending this upstream, don't forget to leave a comment to explain why the projection is not pushed down, why it doesn't matter, and the fact that it would have been pushed down if it mattered.

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -3568,9 +3568,9 @@ Sort: l.sn ASC NULLS LAST
-----ProjectionExec: expr=[zip_code@1 as zip_code, country@2 as country, sn@0 as sn, ts@3 as ts, currency@4 as currency, amount@5 as amount, sum_amount@6 as sum_amount]
+ProjectionExec: expr=[zip_code@1 as zip_code, country@2 as country, sn@0 as sn, ts@3 as ts, currency@4 as currency, amount@5 as amount, sum_amount@6 as sum_amount]
+--SortPreservingMergeExec: [sn@0 ASC NULLS LAST]
+----SortExec: expr=[sn@0 ASC NULLS LAST]
```
When sending this upstream, don't forget to leave a comment to explain why the projection is not pushed down, why it doesn't matter, and the fact that it would have been pushed down if it mattered.

ozankabak commented on datafusion/sqllogictest/test_files/tpch/q2.slt.part:
```diff
@@ -100,89 +100,87 @@ Limit: skip=0, fetch=10
+ProjectionExec: expr=[s_acctbal@5 as s_acctbal, s_name@2 as s_name, n_name@7 as n_name, p_partkey@0 as p_partkey, p_mfgr@1 as p_mfgr, s_address@3 as s_address, s_phone@4 as s_phone, s_comment@6 as s_comment]
+--GlobalLimitExec: skip=0, fetch=10
+----SortPreservingMergeExec: [s_acctbal@5 DESC,n_name@7 ASC NULLS LAST,s_name@2 ASC NULLS LAST,p_partkey@0 ASC NULLS LAST], fetch=10
+------SortExec: TopK(fetch=10), expr=[s_acctbal@5 DESC,n_name@7 ASC NULLS LAST,s_name@2 ASC NULLS LAST,p_partkey@0 ASC NULLS LAST]
```
When sending this upstream, don't forget to leave a comment to explain why the projection is not pushed down, why it doesn't matter, and the fact that it would have been pushed down if it mattered.

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -1647,12 +1649,13 @@ Projection: aggregate_test_100.c9, SUM(aggregate_test_100.c9) PARTITION BY [aggr
-------BoundedWindowAggExec: wdw=[SUM(aggregate_test_100.c9) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 DESC NULLS FIRST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING: Ok(Field { name: "SUM(aggregate_test_100.c9) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 DESC NULLS FIRST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING", data_type: UInt64, nullable: true,  ...
---------SortExec: expr=[c1@0 ASC NULLS LAST,c9@1 DESC]
-----------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/testing/data/csv/aggregate_test_100.csv]]}, projection=[c1, c9], has_header=true
+----ProjectionExec: expr=[c9@1 as c9, SUM(aggregate_test_100.c9) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 DESC NULLS FIRST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@2 as SUM(aggregate_test_100.c9) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 DESC NULLS FIRST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING, SUM(aggregate_test_100.c9) PARTITION BY [aggr ...
```
Why do we have two projections here instead of one? Seems like what we do is more inefficient than the old plan.

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -1603,13 +1605,13 @@ Projection: aggregate_test_100.c9, SUM(aggregate_test_100.c9) ORDER BY [aggregat
---------SortExec: expr=[c1@0 ASC NULLS LAST,c9@1 DESC]
-----------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/testing/data/csv/aggregate_test_100.csv]]}, projection=[c1, c9], has_header=true
-
+----ProjectionExec: expr=[c9@1 as c9, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c1 ASC NULLS LAST, aggregate_test_100.c9 DESC NULLS FIRST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@2 as SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c1 ASC NULLS LAST, aggregate_test_100.c9 DESC NULLS FIRST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING, SUM(aggregate_test_100.c9) PARTITION BY [ag ...
```
Why do we have two projections here instead of one? Seems like what we do is more inefficient than the old plan.

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -1517,23 +1518,24 @@ Projection: SUM(null_cases.c1) ORDER BY [null_cases.c3 ASC NULLS LAST] RANGE BET
 ----------------------TableScan: null_cases projection=[c1, c2, c3]
 physical_plan
-ProjectionExec: expr=[SUM(null_cases.c1) ORDER BY [null_cases.c3 ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@18 as a, SUM(null_cases.c1) ORDER BY [null_cases.c3 ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@18 as b, SUM(null_cases.c1) ORDER BY [null_cases.c3 DESC NULLS FIRST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@3 as c, SUM(null_cases.c1) ORDER BY [null_cases. ...
+ProjectionExec: expr=[SUM(null_cases.c1) ORDER BY [null_cases.c3 ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@16 as a, SUM(null_cases.c1) ORDER BY [null_cases.c3 ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@16 as b, SUM(null_cases.c1) ORDER BY [null_cases.c3 DESC NULLS FIRST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@1 as c, SUM(null_cases.c1) ORDER BY [null_cases. ...
```
Seems like we have 3 projections instead of 2 like in the old plan. AFAICT what we do is more inefficient than the old plan.

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -1435,14 +1435,15 @@ Projection: aggregate_test_100.c9, SUM(aggregate_test_100.c9) ORDER BY [aggregat
 ----------TableScan: aggregate_test_100 projection=[c1, c2, c9]
 physical_plan
-ProjectionExec: expr=[c9@2 as c9, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 ASC NULLS LAST, aggregate_test_100.c1 ASC NULLS LAST, aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@5 as sum1, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 DESC NULLS FIRST, aggregate_test_100.c1 DESC NULLS FIRST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@3 as ...
+ProjectionExec: expr=[c9@0 as c9, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 ASC NULLS LAST, aggregate_test_100.c1 ASC NULLS LAST, aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@3 as sum1, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c9 DESC NULLS FIRST, aggregate_test_100.c1 DESC NULLS FIRST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@1 as ...
```
Why do we have two projections here instead of one? Seems like what we do is more inefficient than the old plan.

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -1693,17 +1696,17 @@ Projection: aggregate_test_100.c3, SUM(aggregate_test_100.c9) ORDER BY [aggregat
 ------------TableScan: aggregate_test_100 projection=[c2, c3, c4, c9]
 physical_plan
-ProjectionExec: expr=[c3@1 as c3, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 DESC NULLS FIRST, aggregate_test_100.c9 DESC NULLS FIRST, aggregate_test_100.c2 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@3 as sum1, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 ASC NULLS LAST, aggregate_test_100.c9 ASC  ...
+ProjectionExec: expr=[c3@0 as c3, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 DESC NULLS FIRST, aggregate_test_100.c9 DESC NULLS FIRST, aggregate_test_100.c2 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@1 as sum1, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 ASC NULLS LAST, aggregate_test_100.c9 ASC  ...
```
Seems like we have 3 projections instead of 2 like in the old plan. AFAICT what we do is more inefficient than the old plan.

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -2137,17 +2138,18 @@ Projection: t1.c9, SUM(t1.c9) PARTITION BY [t1.c1, t1.c2] ORDER BY [t1.c9 ASC NU
 ----------------------TableScan: aggregate_test_100 projection=[c1, c2, c8, c9]
 physical_plan
-ProjectionExec: expr=[c9@1 as c9, SUM(t1.c9) PARTITION BY [t1.c1, t1.c2] ORDER BY [t1.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@4 as sum1, SUM(t1.c9) PARTITION BY [t1.c2, t1.c1_alias] ORDER BY [t1.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@6 as sum2, SUM(t1.c9) PARTITION BY [t1.c1, t1.c2] ORDER BY [t1.c9 ASC NULLS LAST, t1.c8 ASC NULLS LAST] ROWS BETWEEN 1 PRECE ...
+ProjectionExec: expr=[c9@0 as c9, SUM(t1.c9) PARTITION BY [t1.c1, t1.c2] ORDER BY [t1.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@2 as sum1, SUM(t1.c9) PARTITION BY [t1.c2, t1.c1_alias] ORDER BY [t1.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 5 FOLLOWING@4 as sum2, SUM(t1.c9) PARTITION BY [t1.c1, t1.c2] ORDER BY [t1.c9 ASC NULLS LAST, t1.c8 ASC NULLS LAST] ROWS BETWEEN 1 PRECE ...
```
Seems like we have 5 projections instead of 4 like in the old plan. AFAICT what we do is more inefficient than the old plan.

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -2962,16 +2961,17 @@ Projection: annotated_data_infinite2.a, annotated_data_infinite2.b, annotated_da
 ------------------TableScan: annotated_data_infinite2 projection=[a, b, c, d]
 physical_plan
-ProjectionExec: expr=[a@1 as a, b@2 as b, c@3 as c, SUM(annotated_data_infinite2.c) PARTITION BY [annotated_data_infinite2.a, annotated_data_infinite2.d] ORDER BY [annotated_data_infinite2.b ASC NULLS LAST, annotated_data_infinite2.c ASC NULLS LAST] ROWS BETWEEN 2 PRECEDING AND 1 FOLLOWING@9 as sum1, SUM(annotated_data_infinite2.c) PARTITION BY [annotated_data_infinite2.a, annotated_data_infinite ...
+ProjectionExec: expr=[a@0 as a, b@1 as b, c@2 as c, SUM(annotated_data_infinite2.c) PARTITION BY [annotated_data_infinite2.a, annotated_data_infinite2.d] ORDER BY [annotated_data_infinite2.b ASC NULLS LAST, annotated_data_infinite2.c ASC NULLS LAST] ROWS BETWEEN 2 PRECEDING AND 1 FOLLOWING@7 as sum1, SUM(annotated_data_infinite2.c) PARTITION BY [annotated_data_infinite2.a, annotated_data_infinite ...
```
Seems like we have 3 projections instead of 2 like in the old plan. AFAICT what we do is more inefficient than the old plan.

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -4067,13 +4067,14 @@ Projection: aggregate_test_100.c3, SUM(aggregate_test_100.c9) ORDER BY [aggregat
 ----------TableScan: aggregate_test_100 projection=[c3, c4, c9]
 physical_plan
-ProjectionExec: expr=[c3@0 as c3, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@4 as sum2, sum1@3 as sum1]
+ProjectionExec: expr=[c3@0 as c3, SUM(aggregate_test_100.c9) ORDER BY [aggregate_test_100.c3 + aggregate_test_100.c4 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@2 as sum2, sum1@1 as sum1]
```
Seems like we have 3 projections instead of 2 like in the old plan. AFAICT what we do is more inefficient than the old plan.

ozankabak reviewed (commented):
Let's address my comments and then I will do a deep dive in the `optimize_projections.rs` file.

berkaysynnada committed:
Fix projection renaming

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
agg fixes

berkaysynnada committed:
test updates

berkaysynnada committed:
fixing count mismatch

berkaysynnada committed:
Update aggregate_statistics.rs

berkaysynnada committed:
catch different names

berkaysynnada committed:
Fix after merge

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
fix after merge

berkaysynnada committed:
fixing tpch queries

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Update physical_planner.rs

berkaysynnada committed:
Update optimize_projections.rs

berkaysynnada committed:
Remove unnecessary projections

berkaysynnada committed:
Update optimize_projections.rs

berkaysynnada committed:
Fixing all tests

berkaysynnada committed:
Self review part 1

berkaysynnada committed:
Self review part 2

ozankabak committed:
Review

ozankabak commented on datafusion/sqllogictest/test_files/joins.slt:
```diff
@@ -1451,9 +1451,9 @@ logical_plan
-03)----HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(join_t1.t1_id + Int64(11)@3, CAST(join_t2.t2_id AS Int64)@3)], projection=[t1_id@0, t1_name@1, t1_int@2, t2_id@4, t2_name@5, t2_int@6]
+01)CoalesceBatchesExec: target_batch_size=2
+02)--ProjectionExec: expr=[t1_id@0 as t1_id, t1_name@2 as t1_name, t1_int@3 as t1_int, t2_id@4 as t2_id, t2_name@5 as t2_name, t2_int@6 as t2_int, CAST(t1_id@0 AS Int64) + 11 as join_t1.t1_id + Int64(11)]
+03)----HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(join_t1.t1_id + Int64(11)@3, CAST(join_t2.t2_id AS Int64)@3)], projection=[t1_id@0, t1_id@0, t1_name@1, t1_int@2, t2_id@4, t2_name@5, t2_int@6]
```
The projection `t1_id@0` is repeated twice and the projection at the top is missing index `1` (directly skips to 2 from 0). Let's investigate and fix.

ozankabak commented on datafusion/sqllogictest/test_files/joins.slt:
```diff
@@ -1477,9 +1477,9 @@ logical_plan
-03)----HashJoinExec: mode=Partitioned, join_type=Inner, on=[(join_t1.t1_id + Int64(11)@3, CAST(join_t2.t2_id AS Int64)@3)], projection=[t1_id@0, t1_name@1, t1_int@2, t2_id@4, t2_name@5, t2_int@6]
+01)CoalesceBatchesExec: target_batch_size=2
+02)--ProjectionExec: expr=[t1_id@0 as t1_id, t1_name@2 as t1_name, t1_int@3 as t1_int, t2_id@4 as t2_id, t2_name@5 as t2_name, t2_int@6 as t2_int, CAST(t1_id@0 AS Int64) + 11 as join_t1.t1_id + Int64(11)]
+03)----HashJoinExec: mode=Partitioned, join_type=Inner, on=[(join_t1.t1_id + Int64(11)@3, CAST(join_t2.t2_id AS Int64)@3)], projection=[t1_id@0, t1_id@0, t1_name@1, t1_int@2, t2_id@4, t2_name@5, t2_int@6]
```
Same as above.

ozankabak commented on datafusion/sqllogictest/test_files/tpch/q1.slt.part:
```diff
@@ -56,8 +56,9 @@ physical_plan
 07)------------AggregateExec: mode=Partial, gby=[l_returnflag@5 as l_returnflag, l_linestatus@6 as l_linestatus], aggr=[SUM(lineitem.l_quantity), SUM(lineitem.l_extendedprice), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax), AVG(lineitem.l_quantity), AVG(lineitem.l_extendedprice), AVG(linei ...
 08)--------------ProjectionExec: expr=[l_extendedprice@1 * (Some(1),20,0 - l_discount@2) as lineitem.l_extendedprice * (Decimal128(Some(1),20,0) - lineitem.l_discount)Decimal128(Some(1),20,0) - lineitem.l_discountlineitem.l_discountDecimal128(Some(1),20,0)lineitem.l_extendedprice, l_quantity@0 as l_quantity, l_extendedprice@1 as l_extendedprice, l_discount@2 as l_discount, l_tax@3 as l_tax, l_ret ...
 09)----------------CoalesceBatchesExec: target_batch_size=8192
-10)------------------FilterExec: l_shipdate@6 <= 10471
-11)--------------------CsvExec: file_groups={4 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:0..18561749], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:18561749..37123498], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/lineitem.tbl:37123498..55685247], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/tpch/data/line ...
+10)------------------ProjectionExec: expr=[l_quantity@0 as l_quantity, l_extendedprice@1 as l_extendedprice, l_discount@2 as l_discount, l_tax@3 as l_tax, l_returnflag@4 as l_returnflag, l_linestatus@5 as l_linestatus]
```
Why don't we have a single projection here?

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -1517,23 +1517,24 @@ logical_plan
 12)----------------------TableScan: null_cases projection=[c1, c2, c3]
 physical_plan
-01)ProjectionExec: expr=[SUM(null_cases.c1) ORDER BY [null_cases.c3 ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@18 as a, SUM(null_cases.c1) ORDER BY [null_cases.c3 ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@18 as b, SUM(null_cases.c1) ORDER BY [null_cases.c3 DESC NULLS FIRST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@3 as c, SUM(null_cases.c1) ORDER BY [null_cas ...
+01)ProjectionExec: expr=[SUM(null_cases.c1) ORDER BY [null_cases.c3 ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@16 as a, SUM(null_cases.c1) ORDER BY [null_cases.c3 ASC NULLS LAST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@16 as b, SUM(null_cases.c1) ORDER BY [null_cases.c3 DESC NULLS FIRST] RANGE BETWEEN 10 PRECEDING AND 11 FOLLOWING@1 as c, SUM(null_cases.c1) ORDER BY [null_cas ...
```
There is an extra projection here, let's investigate.

ozankabak commented on datafusion/sqllogictest/test_files/tpch/q2.slt.part:
```diff
@@ -100,89 +100,87 @@ logical_plan
-02)--SortPreservingMergeExec: [s_acctbal@0 DESC,n_name@2 ASC NULLS LAST,s_name@1 ASC NULLS LAST,p_partkey@3 ASC NULLS LAST], fetch=10
-03)----SortExec: TopK(fetch=10), expr=[s_acctbal@0 DESC,n_name@2 ASC NULLS LAST,s_name@1 ASC NULLS LAST,p_partkey@3 ASC NULLS LAST]
-04)------ProjectionExec: expr=[s_acctbal@5 as s_acctbal, s_name@2 as s_name, n_name@7 as n_name, p_partkey@0 as p_partkey, p_mfgr@1 as p_mfgr, s_address@3 as s_address, s_phone@4 as s_phone, s_comment@6 as s_comment]
+01)ProjectionExec: expr=[s_acctbal@5 as s_acctbal, s_name@2 as s_name, n_name@7 as n_name, p_partkey@0 as p_partkey, p_mfgr@1 as p_mfgr, s_address@3 as s_address, s_phone@4 as s_phone, s_comment@6 as s_comment]
```
Why is this above the limit?

ozankabak reviewed (commented):
Let's address the reviews and iterate again.

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
before test complete

berkaysynnada committed:
joins are fixed

berkaysynnada committed:
Update optimize_projections.rs

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Ready for review

berkaysynnada committed:
Investigation

Projections are preferred to be on top of
1) GlobalLimitExec
2) LocalLimitExec
3) CoalesceBatchesExec
Only slt tests are modified. Unit test results have not been updated.

berkaysynnada committed:
Revert "Investigation"

This reverts commit c87dbde8fc1a285bb0bbc73e6dc2725da65b8fe1.

ozankabak committed:
Review Part 2

ozankabak committed:
Remove whitespace diffs

berkaysynnada committed:
Do not pushdown projections making calculations

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Update optimize_projections.rs

ozankabak committed:
Update module comments for optimize_projections.rs

ozankabak committed:
Review

berkaysynnada committed:
Unify similar plans

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Update optimize_projections.rs

ozankabak committed:
Refactor ExprMapping

ozankabak committed:
Use error macros

ozankabak committed:
Simplify try_projection_insertion

berkaysynnada committed:
Update optimize_projections.rs

berkaysynnada committed:
Merge branch 'apache_main' into feature/optimize-projections

berkaysynnada committed:
Fix the bug

berkaysynnada committed:
Update cte.slt

berkaysynnada committed:
Update cte.slt
